### DISCUSSION TITLE
[Question] Poor validation performance and garbled inference output when applying NLLB-style language tags to multilingual translation

### DISCUSSION THREAD

**Discussion by hui-li-xf** [eole-nlp/eole#315] [General]
Description I am attempting to replicate the NLLB translation tagging strategy by manually adding language tags to my dataset.

My Approach

Source: Prepended source language tag to the source text (e.g., __eng_Latn__ Hello world).

Target: Prepended target language tag to the target text (e.g., __zho_Hans__ 你好).

The Issue I conducted experiments. While the training loss decreased normally, the validation performance remained poor. Furthermore, during inference, the model generates garbled, repetitive character sequences instead of valid text.

- yaml

<img width="984" height="930" alt="Image" src="https://github.com/user-attachments/assets/ff759671-3f55-4e73-a875-6732d7ceab6e" />

- vocab

<img width="650" height="302" alt="Image" src="https://github.com/user-attachments/assets/4850b159-e2ab-4587-9dfa-cba4f05ca818" />

- train_log

<img width="2616" height="762" alt="Image" src="https://github.com/user-attachments/assets/ca7ae289-eb1a-401c-8590-53e016d981ea" />

**Comment by wildwolf085** [comment]
Something's not quite right with your *YAML* setup. Here's what to try...
The important is to change `onmt-tokenize` to `sentencepiece`.

```yaml
transforms: [sentencepiece, filtertoolong]
# transforms configuration
transforms_configs:
  sentencepiece:
    src_subword_model: spm_bpe_25600.model
    tgt_subword_model: spm_bpe_25600.model

  filtertoolong:
    src_seq_length: 512
    tgt_seq_length: 512
```

**Reply by hui-li-xf** [reply]
I referred to the configuration on the official website.https://github.com/eole-nlp/eole/blob/main/recipes/wmt17/wmt17_ende_spm_onmt_tokenize.yaml
onmt-tokenize seems to be working fine.

**Comment by vince62s** [comment]
here is an example of yaml for what you are trying to do:

```
seed: 1234
share_vocab: true
#src_vocab: "/en-de/params/llama3/vocab.txt"
#src_vocab_size: 128256
src_vocab: "/en-de/params/ende.vocab2"
#src_vocab_size: 24936

src_words_min_frequency: 1
vocab_size_multiple: 8
report_every: 100
skip_empty_level: silent
valid_metrics: ["BLEU"]
scoring_debug: True

# transforms config
transforms: [onmt_tokenize, filtertoolong]
transforms_configs:
  onmt_tokenize:
    #### Subword
    src_subword_type: bpe
    src_subword_model: "/en-de/params/subwords.en_de.bpe"
    src_onmttok_kwargs: {"mode": "aggressive", "joiner_annotate": True, "preserve_placeholders": True, "case_markup": True, "soft_case_regions": True, "preserve_segmented_tokens": True, "segment_case": True, "segment_numbers": True, "segment_alphabet_change": True}
    tgt_subword_type: bpe
    tgt_subword_model: "/en-de/params/subwords.en_de.bpe"
    tgt_onmttok_kwargs: {"mode": "aggressive", "joiner_annotate": True, "preserve_placeholders": True, "case_markup": True, "soft_case_regions": True, "preserve_segmented_tokens": True, "segment_case": True, "segment_numbers": True, "segment_alphabet_change": True}
  filtertoolong:
    src_seq_length: 1024
    tgt_seq_length: 1024

# Corpus opts:
data:
    nc-v18:
# filtered 362408
        path_tgt: "/en-de/news-commentary/news-commentary-v18.de-en.de"
        path_src: "/en-de/news-commentary/news-commentary-v18.de-en.en"
        transforms: [onmt_tokenize, docify, prefix, filtertoolong]
        weight: 1
        src_prefix: eng_Latn
        tgt_prefix: deu_Latn
    europarl:
# filtered 1062568
        path_tgt: "/en-de/europarl/europarl-v10.de-en.de"
        path_src: "/en-de/europarl/europarl-v10.de-en.en"
        transforms: [onmt_tokenize, docify, prefix, filtertoolong]
        weight: 3
        src_prefix: eng_Latn
        tgt_prefix: deu_Latn
    cc-matrix:
# filtered 34164475
        path_src: "/en-de/cc-matrix/cc-matrix-ende.en"
        path_tgt: "/en-de/cc-matrix/cc-matrix-ende.de"
        transforms: [onmt_tokenize, prefix, filtertoolong]
        weight: 100
        src_prefix: eng_Latn
        tgt_prefix: deu_Latn
    valid:
        path_src: "/en-de/testsets/newstest2014-origlang-en.en"
        path_tgt: "/en-de/testsets/newstest2014-origlang-en.de"
        transforms: [onmt_tokenize, prefix]
        src_prefix: eng_Latn
        tgt_prefix: deu_Latn

log_file: "/en-de/runs/base_educational/trf_trf_6-6-8-512-2048/training.log"

training:
    # General opts
    #update_vocab: true
    #train_from: "/en-de/runs/base_educational/trf_trf_6-6-8-512-2048"
    #reset_optim: states
    model_path: "/en-de/runs/base_educational/trf_trf_6-6-8-512-2048"
    keep_checkpoint: 50
    save_checkpoint_steps: 10000
    average_decay: 0.0 #0.0005
    train_steps: 100000
    valid_steps: 10000

    # Batching
    bucket_size: 262144
    num_workers: 4
    prefetch_factor: 400
    world_size: 1
    gpu_ranks: [0]
    parallel_mode: "data_parallel"
    batch_type: "tokens"
    batch_size: 32768
    valid_batch_size: 4096
    batch_size_multiple: 1
    accum_count: [1, 1, 1]
    accum_steps: [0, 15000, 30000]

    # Optimization
    compute_dtype: "fp16"
    use_amp: true
    optim: "adamw"
    weight_decay: 0.01
    reset_optim: "none"
    learning_rate: 1
    warmup_steps: 6000
    decay_method: "noam"
    adam_beta2: 0.998
    adam_eps: 1e-8
    max_grad_norm: 1.0
    label_smoothing: 0.1
    param_init_method: "normal" #"xavier_uniform"
    param_init: 0.02
    normalization: "tokens"

    dropout_steps: [0, 15000, 30000]
    dropout: [0.1, 0.1, 0.1]
    attention_dropout: [0.0, 0.0, 0.0]
    estim_loss_lambda_steps: [0, 10000, 20000, 50000, 100000, 150000, 160000, 170000]
    estim_loss_lambda: [1.0, 2.0, 5.0, 10.0, 20.0, 40.0, 80.0, 200.0]
    score_threshold: 0.65

    freeze_decoder: false
    freeze_encoder: false

model:
    architecture: "transformer"
    encoder:
        layers: 6
    decoder:
        layers: 6
    hidden_size: 512
    heads: 8
    transformer_ff: 2048
    add_qkvbias: false
    add_ffnbias: true
    mlp_activation_fn: gelu
    add_estimator: false
    share_decoder_embeddings: true
    share_embeddings: true
    layer_norm: standard
    norm_eps: 1e-5
    embeddings:
        word_vec_size: 512
        position_encoding_type: "SinusoidalInterleaved"
        freeze_word_vecs_dec: false
  ```

**Reply by hui-li-xf** [reply]
Do `src_prefix` and `tgt_prefix` always need to be set in the training data? If I'm training in multiple languages, it seems I can't directly add `src_prefix` and `tgt_prefix` to the `valid` function because it contains multiple languages.

**Comment by vince62s** [comment]
it is much easier to do so with one target language set in the prefix transform.
If you want to handle manually your tags in source / target, you will need to play with a specific file with all target prefix for each line, plus the tgt_file_prefix flag. More cumbersome.
I think you can "test" outside the valid step if your model does fine for other languages.

**Reply by hui-li-xf** [reply]
I'm currently adding relevant language labels to the beginning of each row in the training data. However, even when I use the model to perform inference tests independently, it fails to infer correctly. The reasoning result is garbled text.

**Comment by hui-li-xf** [comment]
I tried adding src_prefix and tgt_prefix, but I still couldn't get the multilingual translation to work properly. Is it because teacher focusing is not supported?
data:
    en-zh:
        path_tgt: xxxx
        path_src: xxxx
        transforms: [onmt_tokenize, docify, prefix, filtertoolong]
        weight: 1
        src_prefix: eng_Latn
        tgt_prefix: deu_Latn
    en-ru:
        path_tgt: xxxx
        path_src: xxxx
        transforms: [onmt_tokenize, docify, prefix, filtertoolong]
        weight: 3
        src_prefix: eng_Latn
        tgt_prefix: zho_Hans

**Comment by vince62s** [comment]
post your full yaml file if you want help.
